### PR TITLE
Add HTTP Basic Authorisation to search.

### PR DESCRIPTION
**Pull request by mhilton**
If a search call is performed with valid Basic Authorisation credentials then
the search will be in admin mode and will not check ACLs.

### CODE REVIEW

**Comment by mhilton** [comment]
## QA Instructions

Add some ACLs on charms in the charmstore and then perform a search on those charms prividing valid admin credentials and make sure they're shown.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/712/
Test PASSed.

File: internal/v4/api_test.go [hunk 1]
```diff
@@ -78,7 +79,7 @@ func (s *APISuite) SetUpTest(c *gc.C) {
 
 func newServer(c *gc.C, session *mgo.Session, si *charmstore.SearchIndex, config charmstore.ServerParams) (http.Handler, *charmstore.Store) {
 	db := session.DB("charmstore")
-	store, err := charmstore.NewStore(db, si, nil)
+	store, err := charmstore.NewStore(db, si, &bakery.NewServiceParams{})
```
Comment (rogpeppe) [inline internal/v4/api_test.go:82]: I wonder if it would be better if NewStore took a value rather than a pointer.
If we keep it as a pointer, why is this change necessary?

File: internal/v4/auth.go [hunk 2]
```diff
@@ -78,6 +79,32 @@ func (h *Handler) authorize(req *http.Request, acl []string) error {
 	return httpbakery.NewDischargeRequiredError(m, verr)
 }
 
+// checkRequest performs an httpbakery.CheckRequest and then filters out any
+// declared caveat attributes that were not requested.
+func (h *Handler) checkRequest(req *http.Request) (map[string]string, error) {
```
Comment (rogpeppe) [inline internal/v4/auth.go]: Let's make this return a struct:

```
type authDeclaredAttrs struct {
    Username string
    Groups []string
}
```

That way we know exactly what we've got, and this function
becomes simpler too:

```
return &authDeclaredAttrs{
    Username: attrs[usernameAttr],
    Groups: strings.Fields(attrs[groupsAttr]),
}
```

File: internal/v4/auth.go [hunk 3]
```diff
@@ -78,6 +79,32 @@ func (h *Handler) authorize(req *http.Request, acl []string) error {
...
+	attrs, err := httpbakery.CheckRequest(h.store.Bakery, req, nil, checkers.New())
+	if err != nil {
+		return nil, errgo.Mask(err, errgo.Any)
+	}
+	for k := range attrs {
```
Comment (frankban) [inline internal/v4/auth.go]: A comment about why we are doing this sanity check?

File: internal/v4/search_test.go [hunk 4]
```diff
@@ -51,39 +48,27 @@ var exportTestBundles = map[string]string{
...
-		},
-		URL:      storeURL("trusty/riak-67/meta/perm/read"),
-		Username: serverParams.AuthUsername,
-		Password: serverParams.AuthPassword,
-	})
-	err := s.ES.RefreshIndex(s.TestIndex)
+	// hide the riak charm
+	err := s.store.DB.BaseEntities().UpdateId(
+		charm.MustParseReference("cs:riak"),
+		bson.D{{"$set", map[string]map[string][]string{
```
Comment (frankban) [inline internal/v4/search_test.go]: Use mongodoc.ACL instead?

**Comment by frankban** [comment]
:+1: with a couple of minor suggestions. Thank you!

File: internal/v4/search.go [hunk 5]
```diff
@@ -137,17 +135,24 @@ func parseSearchParams(req *http.Request) (charmstore.SearchParams, error) {
 	return sp, nil
 }
 
+// groupsFromRequest finds the Read ACL users and groups that should be added to the
+// SearchParams.Groups list. It will also determine if the requestee has full admin rights.
 func (h *Handler) groupsFromRequest(sp *charmstore.SearchParams, req *http.Request) {
```
Comment (rogpeppe) [inline internal/v4/search.go]: I kind of feel that this doesn't need to be tied so closely to SearchParams.
It could be groupsFromRequest(req *http.Request) (isAdmin bool, groups []string, err error)
and which is closer to the contract implied by the name.

But when you look at it that way, it's not really much different from checkRequest,
which takes a request and returns some user info.

So perhaps better still might be to change checkRequest to check for admin credentials,
and add Admin to authDeclaredAttrs (authInfo might be a better name for that).

Then checkBasicAuth only needs to be called in one place.

**Comment by rogpeppe** [comment]
Really nice stuff, thanks. I have a couple of suggestions above. What do you think?

File: internal/v4/api_test.go (hunk 1 above)
Comment (mhilton) [inline internal/v4/api_test.go:82]: If the pointer is nil then a bakery is not created, where as if it is empty parameters it is created with default configuration. My tests require the latter.

File: internal/v4/search_test.go (hunk 4 above)
Comment (rogpeppe) [inline internal/v4/search_test.go]: +1

File: internal/v4/auth.go (hunk 2 above)
Comment (mhilton) [inline internal/v4/auth.go]: done. I like that.

File: internal/v4/search_test.go (hunk 4 above)
Comment (mhilton) [inline internal/v4/search_test.go]: done. thanks, code's much simpler.

File: internal/v4/search.go (hunk 5 above)
Comment (mhilton) [inline internal/v4/search.go]: done. This has really simplified things.

File: internal/v4/auth.go (hunk 3 above)
Comment (mhilton) [inline internal/v4/auth.go]: This has changed quite a bit, but I've expanded the comment.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/714/
Test PASSed.

File: internal/v4/auth.go [hunk 6]
```diff
@@ -82,6 +72,34 @@ func (h *Handler) authorize(req *http.Request, acl []string) error {
 	return httpbakery.NewDischargeRequiredError(m, cookiePath, verr)
 }
 
+// checkRequest checks for any authorisation tokens in the request either HTTP
```
Comment (rogpeppe) [inline internal/v4/auth.go]: I'm not sure we need to describe all the implementation in this comment.
I'd limit it to describing the purpose of the function and how to use it.

```
// checkRequest checks the authorization provided by the
// given request. On success, it reports whether the request has admin
// rights; without admin rights it returns information on the authenticated
// user.
```

This is slightly more awkward to phrase because Admin isn't part of
the struct.

File: internal/v4/auth.go [hunk 7]
```diff
@@ -45,29 +45,19 @@ func (h *Handler) authorize(req *http.Request, acl []string) error {
...
-		return errgo.WithCausef(err, params.ErrUnauthorized, "authentication failed")
-	}
-
-	// No basic auth presented: use macaroon third-party caveat verification.
-	attrs, verr := httpbakery.CheckRequest(h.store.Bakery, req, nil, checkers.New())
+	admin, attrs, verr := h.checkRequest(req)
 	if verr == nil {
-		logger.Infof("authenticated with attrs %q", attrs)
+		logger.Infof("authenticated with admin: %q, attrs %q", admin, attrs)
+		if admin {
```
Comment (rogpeppe) [inline internal/v4/auth.go]: This is another place where having Admin in the struct returned by checkRequest
would make things simpler (the check could be done in checkACLMembership)

File: internal/v4/search.go [hunk 8]
```diff
@@ -28,7 +25,14 @@ func (h *Handler) serveSearch(_ http.Header, req *http.Request) (interface{}, er
 	if err != nil {
 		return "", err
 	}
-	h.groupsFromRequest(&sp, req)
+	admin, attrs, err := h.checkRequest(req)
```
Comment (rogpeppe) [inline internal/v4/search.go]: check error here.

there should probably be a test for that.

**Comment by rogpeppe** [comment]
LGTM with the error check fixed, and one suggestion about putting Admin inside the struct (if you did, authInfo or clientAuthorizationInfo would probably be a better name for it). Thanks!

File: internal/v4/auth.go (hunk 6 above)
Comment (mhilton) [inline internal/v4/auth.go]: I've updated this so admin is part of the struct.

File: internal/v4/auth.go (hunk 7 above)
Comment (mhilton) [inline internal/v4/auth.go]: done

File: internal/v4/search.go (hunk 8 above)
Comment (mhilton) [inline internal/v4/search.go]: I don't actually want to error here, just carry on as if there is no logged in user, which there probably isn't, I have added a comment explaining this.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/715/
Test PASSed.

File: internal/v4/auth.go [hunk 9]
```diff
@@ -119,16 +131,26 @@ const (
 	groupsAttr   = "groups"
 )
 
-func (h *Handler) checkACLMembership(attrs map[string]string, acl []string) error {
-	user := attrs[usernameAttr]
-	if user == "" {
+// authorizartion conatains authorization information extracted from the request.
```
Comment (rogpeppe) [inline internal/v4/auth.go]: s/authorizartion/authorization/
s/the request/an HTTP request/ ?

File: internal/v4/search.go [hunk 10]
```diff
@@ -28,7 +25,14 @@ func (h *Handler) serveSearch(_ http.Header, req *http.Request) (interface{}, er
 	if err != nil {
 		return "", err
 	}
-	h.groupsFromRequest(&sp, req)
+	// Ignore the error from checkRequest and carry on assuming no authorization was
+	// sent.
+	auth, _ := h.checkRequest(req)
```
Comment (rogpeppe) [inline internal/v4/search.go]: Seems like a good approach in general, but I'd
be inclined to log the error (Info level?)
e.g.
log.Infof("authorization failed on search request, granting no privileges: %v", err)

?

**Comment by rogpeppe** [comment]
LGTM with one trivial suggestion, thanks!

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/716/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/717/
Test PASSed.

**Comment by mhilton** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge
